src/lib/operators/join_sort_merge.cpp

```diff
@@ -132,6 +132,9 @@ class JoinSortMerge::JoinSortMergeImpl : public AbstractReadOnlyOperatorImpl {
   std::map<RowID, bool> _left_row_ids_emitted{};
   std::map<RowID, bool> _right_row_ids_emitted{};
 
+  std::vector<std::map<RowID, bool>> _left_row_ids_emitted_per_chunks;
```

Comment:
Does this fix #1643 and does it allow us to tackle #2014 next?

Reply:
I wasn't able to reproduce the error from #1643 (master) on my machine 😞 . Sadly I found another error on master. If we use SMJ as the default join and are running the test suit the test `SQLiteTestRunnerUnencoded/SQLiteTestRunner.CompareToSQLite/Line162WithUnencoded` fails with the exception: `C++ exception with description "src/lib/operators/alias_operator.cpp:85 Chunk is sorted by an invalid column ID." thrown in the test body.`.

Reply:
Feel free to move this into a new issue.

Reply:
Also, there is a good chance that #1643 is easier to reproduce on machines with many cores.

Reply:
I create a new Issue for the error #2321 . Still, I have some problems reproducing #1643 . So just to double-check:  

* I use clang by setting up CMake/Ninja with: `CC=/usr/bin/clang CXX=/usr/bin/clang++ cmake -DENABLE_NUMA_SUPPORT=Off -G Ninja ..`
* In https://github.com/hyrise/hyrise/blob/37d36653e0c6c2977f9b188381aa99cae2e2edc9/src/test/lib/sql/sqlite_testrunner/sqlite_testrunner.cpp#L56 I add the lines `Hyrise::get().topology.use_numa_topology(); Hyrise::get().set_scheduler(std::make_shared<NodeQueueScheduler>());`
* Then I build the `hyriseTest` executable and rin `./hyriseTest --gtest_filter=SQLiteTestRunnerUnencoded/SQLiteTestRunner.CompareToSQLite/Line192WithUnencoded` (https://github.com/hyrise/hyrise/blob/8827a25805e866ee7cc393f1b55698ddf6ef4d23/resources/test_data/sqlite_testrunner_queries.sql#L192)

I run the test on Aptera with 80 cores (If I check htop while running the test I can see that I am using all the cores).
Am I doing something wrong, or is it just bad luck? 😞

Reply:
> I use clang by setting up CMake/Ninja with: CC=/usr/bin/clang CXX=/usr/bin/clang++ cmake -DENABLE_NUMA_SUPPORT=Off -G Ninja ..

Any reason you turn off NUMA? Also, I believe the canonical cmake method is `-DCMAKE_C(XX)_COMPILER=clang(++)`. Neither of those *should* make a difference though.

Maybe you could try running the test with `--gtest_repeat=-1` and simply hammer the test until it breaks.

Otherwise, have you understood the issue enough to write a query on the TPC-H data that could trigger the bug? Maybe we are too fast by now for the bug to occur on the sqlite data.

Reply:
> Any reason you turn off NUMA?  

Not really, that's my mistake. 

> Otherwise, have you understood the issue enough to write a query on the TPC-H data that could trigger the bug? Maybe we are too fast by now for the bug to occur on the sqlite data.

I already wrote a little python script awhile ago with one query that should trigger the bug on a hyrise server that utilizes multiple cores. But it is not using the TPC-H data though. I could try to adjust the query to use the TPC-H data. The query was: 

```sql
SELECT * FROM table_1 LEFT JOIN table_2 ON table_1.a > table_2.a AND table_1.b >= table_2.b AND table_1.c < table_2.c
```

<details><summary>script</summary>
<p>

```python
#!/usr/bin/env python3

from psycopg2 import connect
from random import randrange
from multiprocessing import Process

def insert_task(host, port, rows, id):
    conn = connect(host=host, port=port)
    conn.set_session(autocommit=True)
    cur = conn.cursor()
    number_range = 5

    for _ in range(rows):
        cur.execute(f"INSERT INTO table_1 (a, b, c) VALUES ({randrange(number_range)}, {randrange(number_range)}, {randrange(number_range)});")
        cur.execute(f"INSERT INTO table_2 (a, b, c) VALUES ({randrange(number_range)}, {randrange(number_range)}, {randrange(number_range)});")

    print(f"-> Worker {id} done")

    cur.close()
    conn.close()

def load_table(host, port):
    conn = connect(host=host, port=port)
    conn.set_session(autocommit=True)
    cur = conn.cursor()

    cur.execute("CREATE TABLE table_1 (a int, b int, c int);")
    cur.execute("CREATE TABLE table_2 (a int, b int, c int);")

    cur.close()
    conn.close()

def insert(host, port, rows, workers):
    worker_processes = []
    for worker_id in range(workers):
        process = Process(target=insert_task, args=(host, port, rows, worker_id,))
        process.start()
        worker_processes.append(process)

    print("Waiting for workers to finish:")
    for process in worker_processes:
        process.join()
        process.terminate()


def execute_sql(host, port, sql, times):
    conn = connect(host=host, port=port)
    conn.set_session(autocommit=True)
    cur = conn.cursor()
    results = []

    print("Executing sql query:")
    for i in  range(times):
        print(f"-> Execute number: {i}")
        cur.execute(sql)
        result = cur.fetchall()
        results.append(result)
        if i > 0 and results[0] != result:
            print("Results do not match")
            return

    # for result in results:
    #     if results[0] != result:
    #         print("Results do not match")

def main():
    host = "0.0.0.0"
    port = "5432"
    rows_per_worker = 100_000
    workers = 10
    sql = "SELECT * FROM table_1 LEFT JOIN table_2 ON table_1.a > table_2.a AND table_1.b >= table_2.b AND table_1.c < table_2.c;"

    load_table(host, port)
    insert(host, port, rows_per_worker, workers)
    execute_sql(host, port, sql, 6)

if __name__ == '__main__':
    main()
```

</p>
</details>

> Maybe you could try running the test with --gtest_repeat=-1 and simply hammer the test until it breaks.

I think I will try this method for now.

Reply:
It's not about the TPC-H data in particular, I just wonder if we should use more data to trigger the bug more reproducibly.

Reply:
I am working on verifying that we no longer crash here. More or less by chance, we made it really hard to trigger the bug with #2004. This PR introduced checking for the size of jobs and only spawned tasks when the jobs have been big enough.
With that, we never had sufficiently large joins in the tests.

Still not super easy to trigger because once we have larger joins, we also use radix clustering ~that~ which kind of works against larger clusters. As soon as we enforce several radix clusters *and* always spawn jobs, it more or less crashes immediately.
But the point that is not easy to trigger obviously does not mean that we will never see it happen.

I'll ping you here, once I confirmed that it does not longer crash with this PR.

Reply:
> we made it really hard to trigger the bug with #2004

Good catch!

Reply:
So it ran for 2 hours without any problems.

I wanted to push and revert the code to have a reproducible example in the PR. But I think I failed. Sorry for that.

Reply:
I try to revert it, but it kind of failed. Is it okay if we somehow force push 50c8249 and then commit the code (that tests the bug) without any merges in it, or even better in a separate branch where this branch is the base? We could refer to this branch here.

Reply:
> even better in a separate branch where this branch is the base? We could refer to this branch here.

That would make it harder to use `git blame` later to understand the history of the PR.

I don't see too much wrong with force-pushing into a PR. If you are the only developer on that branch, it's fine in virtually any case; otherwise, just sync with the other people who use this branch.